File: insertion/cpp/insertionPessoa.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <time.h>
#include <string>
#include <sstream>
using namespace std;

class Pessoa
{
public:
    int codigo;
    string name;
    Pessoa(){};
    Pessoa(int cod, string nome)
    {
        codigo = cod;
        name = nome;
    }
    void PrintPessoa()
    {
        cout << codigo << ' ' << name << endl;
    }
    bool operator<(Pessoa b)
    {
        return codigo < b.codigo;
    }
};

int main(int argc, char *argv[2])
{
    ifstream input("pessoa 1.txt");
    ofstream output("./insCppPessoa.txt", ios_base::app);
    vector<Pessoa> vetor;
    unsigned long comp = 0, swap = 0;
    clock_t inicio, fim;

    for (int i = 0; i < atoi(argv[1]); i++)
    {
        string linha;

        getline(input, linha);
        stringstream ss(linha);
        string id_str, nome;

        getline(ss, id_str, '\t');
        getline(ss, nome, '\t');

        int id = stoi(id_str);
        vetor.push_back(Pessoa(id, nome));
    }

    // for (int i = 0; i < size(vetor) - 1; i++)
    // {
    //     cout << vetor[i].codigo << " " << vetor[i].name << endl;
    // }
    // cout << endl;

    inicio = clock();

    for (int i = 0; i < size(vetor) - 1; i++)
    {
        Pessoa aux;
        for (int j = i + 1; j > 0; j--)
        {
            comp++;
            if (vetor[j] < vetor[j - 1])
            {
                aux = vetor[j];
                vetor[j] = vetor[j - 1];
                vetor[j - 1] = aux;
                swap++;
            }
            else
                break;
        }
    }

    fim = clock();
    
    // for (int i = 0; i < size(vetor) - 1; i++)
    // {
    //     cout << vetor[i].codigo << " " << vetor[i].name << endl;
    // }

    output << argv[1] << "," << (float)(((fim - inicio) + 0.0) / CLOCKS_PER_SEC) << "," << comp << "," << swap << "\n";

    return 0;
}

File: insertion/golang/insertionPessoa.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Pessoa struct {
	codigo int
	name   string
}

func main() {
	input, err := os.Open("pessoa 1.txt")
	if err != nil {
		panic(err)
	}
	defer input.Close()
	output, err := os.OpenFile("./insGoPessoa.txt", os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		panic(err)
	}
	defer input.Close()

	vetor := make([]Pessoa, 0)
	var comp, swap int

	num, err := strconv.Atoi(os.Args[1])

	scanner := bufio.NewScanner(input)
	for i := 0; i < num; i++ {

		scanner.Scan()
		linha := scanner.Text()
		valores := strings.Split(linha, "\t")
		id, _ := strconv.Atoi(valores[0])
		aux := Pessoa{id, valores[1]}
		vetor = append(vetor, aux)
	}

	// for _, v := range vetor {
	// 	fmt.Printf("%d %s, ", v.codigo, v.name)
	// }
	// println("\nsort\n")

	inicio := time.Now()
	for i := 0; i < len(vetor)-1; i++ {
		for j := i + 1; j > 0; j-- {
			comp++
			if vetor[j].codigo < vetor[j-1].codigo {
				var aux1 Pessoa
				aux1 = vetor[j]
				vetor[j] = vetor[j-1]
				vetor[j-1] = aux1
				swap++
			} else {
				break
			}
		}
	}
	fim := time.Now()

	// for _, v := range vetor {
	// 	fmt.Printf("%d %s\n", v.codigo, v.name)
	// }

	fmt.Fprintf(output, "%s,%v,%d,%d\n", os.Args[1], float64(fim.Sub(inicio).Seconds()), comp, swap)
}
